lib/SILOptimizer/Transforms/SILMem2Reg.cpp

```diff
@@ -957,50 +1056,125 @@ void StackAllocationPromoter::fixBranchesAndUses(BlockSetVector &phiBlocks,
-    bool isFunctionExit = bb->getSuccessorBlocks().empty();
-    auto successorsHaveLiveInValues = [&]() -> bool {
-      return llvm::all_of(bb->getSuccessorBlocks(), [&](auto *successor) {
-        return getLiveInValues(phiBlocks, successor);
-      });
-    };
-    auto hadDeallocStack = [&]() -> bool { return dsiBlocks.contains(bb); };
-    if (isFunctionExit || !successorsHaveLiveInValues() || hadDeallocStack()) {
-      auto values = getLiveOutValues(phiBlocks, bb);
-      if (!values)
+  // We need to separately consider and visit incoming unopened borrow scopes
+  // and outgoing unclosed borrow scopes.  The reason is that a walk should stop
+  // on any path where it encounters an incoming unopened borrow scope but that
+  // should _NOT_ count as a visit of outgoing unclosed borrow scopes.
+  //
+  // Without this distinction, a case like the following wouldn't be visited
+  // properly:
+  //
+  //     bb1:
+  //       %addr = alloc_stack
+  //       store %value to [init] %addr
+  //       br bb2
+  //     bb2:
+  //       %value_2 = load [take] %addr
+  //       store %value_2 to [init] %addr
+  //       br bb3
+  //     bb3:
+  //       destroy_addr %addr
+  //       dealloc_stack %addr
+  //       %r = tuple ()
+  //       return %r
+  //
+  // Both bb1 and bb2 have unendedLexicalLifetimes.  Suppose that we visited bb1
+  // first.  We would see that it didn't have an incoming unopened borrow scope
+  // (already, we can tell something is amiss that we're considering this) and
+  // then add bb2 to the worklist--except it's already there.  Next we would
+  // visit bb2.  We would see that it had an incoming unopened borrow scope so
+  // we would close it.  And then we'd be done.  In particular, we'd leave the
+  // scope that opens in bb2 unclosed.
+  //
+  // The root cause here is that it's important to stop walking when we hit a
+  // scope close.  Otherwise, we could keep walking down to blocks which don't
+  // have live-in or live-out values.
+  //
+  // Visiting the incoming and outgoing edges works as follows in the above
+  // example:  The worklist is initialiized with {(bb1, ::Out), (bb2, ::Out)}.
+  // When visiting (bb1, ::Out), we see that bb1 is neither unreachable nor
+  // has any successors without live-in values.  So we add (bb2, ::In) to the
+  // worklist.  Next, we visit (bb2, ::Out).  We see that it _also_ doesn't have
+  // an unreachable terminator or any successors without live-in values, so we
+  // push (bb3, ::In).  Next, we visit (bb2, ::In).  We see that it _does_ have
+  // an incoming unopened borrow scope, so we close it and stop.  Finally, we
+  // visit (bb3, ::Out).  We see that it too has an incoming unopened borrow
+  // scope so we close it and stop.
+  enum class LiveValuesKind : uint8_t { In, Out };
+
+  using ScopeEndPosition = std::pair<SILBasicBlock *, LiveValuesKind>;
+
+  DAGNodeWorklist<ScopeEndPosition, 16, llvm::SmallSet<ScopeEndPosition, 16>>
+      worklist;
```

Comment:
1. Can you use PointerIntPair? Then you don't need any special worklist template parameters.

2. Better yet, it seems that the code would be a lot easier to understand if had two worklists, one for each purpose. No need for a `LiveValuesKind` enum

3. Better yet, can you just handle the `LiveValuesKind::In` case before pushing that block onto the regular worklist. Then you just have a normal forward CFG traversal, with the usual worklist. You just have two separate points in the traversal where you handle the In vs. Out case

Reply:
As discussed, went with 1 because I think it most clearly shows how the walk proceeds at each step.